---
title: "Downstream_Analysis"
author: "Jinny Chun"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: united
    df_print: kable
  pdf_document: default
  
params: 
  data: "mouse_colon"
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# ST-Analysis

This is an RMarkdown script to analyze spatial transcriptomics (ST) data.

1. Normalization of count-data 
2. Dimensionality reduction and clustering
3. Visualization of features (i.e. genes,clusters, etc.) on spots under the tissue 
4. Differential gene expression analysis

## Library Loading

```{r, load required packages,warning = FALSE, results='hide'}
library(Seurat)
library(STutility)
library(ggplot2)
library(patchwork)
library(dplyr)
library(Matrix)
library("htmltools")
library("vembedr")
```


## Load Data

Import the infotable including paths to the count matrix, coordinate metadata, json file, and images and create the object

```{r load data}
infoTable <- read.table("C:/Users/jinny/Rstudio/DP/info_table.csv", sep=";", header=T, stringsAsFactors = F)
infoTable$sample_id <- paste0("sample_", 1:nrow(infoTable))
```

```{r create_seurat}

se <- InputFromTable(infotable = infoTable)
```

```{r loadimages}

se <- LoadImages(se, time.resolve = FALSE, xdim = 1e3)

```

```{r plot_images, fig.width=12, fig.height=6, out.width="100%"}
ImagePlot(se, method = "raster")
```


### Quality controls for the created object

nFeature_RNA is the total number of genes in each cell that have at least one UMI count. nCount_RNA is the total number of molecules detected within a cell. Low nFeature_RNA for a cell indicates that it may be dead/dying or an empty droplet. High nCount_RNA and/or nFeature_RNA indicates that the "cell" may in fact be a doublet (or multiplet). 

```{r plot_features, fig.width=7, fig.height=7}
for (i in 1:nrow(infoTable)){print(ST.FeaturePlot(se, features = c("nFeature_RNA"),indices = i, pt.size = 2.5))}

```

```{r qc_violin, fig.width=8, fig.height=6}

se$sample_id <- paste0("sample_", GetStaffli(se)@meta.data$sample)
se <- SetIdent(se, value = "day")
p <- VlnPlot(se, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2, pt.size = 0.5)
p

```

## Normalization for samples
Load all samples into the environment to run first read count normalizations,  
check for data quality to exclude samples for downstream analysis


```{r analysis_wf}

se <- se %>% SCTransform(return.only.var.genes = FALSE, variable.features.n = NULL, variable.features.rv.th = 1.1)

```


Print 10 most variable genes in the data

```{r, variable genes}
head(VariableFeatures(se),10)
```

```{r, spatial plot with variable gene expression}


for (i in 1:nrow(infoTable)){print(ST.FeaturePlot(se, features = c(head(VariableFeatures(se),2)), split.labels = F, indices = i,pt.size = 2.5))}


```

```{r, sptial plot with sample images}

for (i in 1:nrow(infoTable)){print(FeatureOverlay(se, features = c(head(VariableFeatures(se),2)), spots = rownames(subset(se[[]])),sampleids = i, pt.alpha = 0.8, pt.size = 2.5))}

```


## Dimensionality Reduction & Clustering

```{r dim.cluster}
se <- RunPCA(se, assay = "SCT", verbose = FALSE)
se <- RunUMAP(se, dims = 1:30, verbose = FALSE)
se <- FindNeighbors(se, reduction = "pca", dims = 1:30)
se <- FindClusters(se, verbose = FALSE)
#se <- RunUMAP(se, reduction = "pca", dims = 1:30)
```



Make feature plots

```{r, feature plots}
se <- SetIdent(se, value = "seurat_clusters")

for (i in 1:nrow(infoTable)){print(ST.FeaturePlot(se, features = "seurat_clusters", split.labels = F, indices = i,pt.size = 2.5))}

```

Visualize clusters on Umap
 
```{r Umap}
DimPlot(se, pt.size = 1 , label = T)
```

Overlay clusters on the images

```{r, spatial plots with slide images}

for (i in 1:nrow(infoTable)){print(FeatureOverlay(se, features = "seurat_clusters", spots = rownames(subset(se[[]])),sampleids = i, pt.alpha = 0.8, pt.size = 3.2))}

```

## Differential gene expression analysis

To identify characteristic markers genes of the previously identified clusters differential genes expression is performed. To specify enriched genes in the individual clusters we set the parameters to only include markers with a positive log thereshold


```{r, find all markers}
#Find the markers for all clusters
se.markers <- FindAllMarkers(se)  
head(se.markers)
inceasing <- se.markers %>% top_n(n = 100, wt = avg_log2FC)
```

Visualization of most variable genes for the identified clusters can be performed by a heatmap. How many genes one wishes to depict can be changed as described below. For the manuscript we decided to depict the 5 genes with the highest logFC genes of each cluster.

```{r, heatmap}

#top 20 genes
top20de <- se.markers %>% group_by(cluster) %>% top_n(n = 20, wt = avg_log2FC)
#top 10 genes
top10de <- se.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)
#top 5 genes
top5de <- se.markers %>% group_by(cluster) %>% top_n(n = 5, wt = avg_log2FC)

top3de <- se.markers %>% group_by(cluster) %>% top_n(n = 3, wt = avg_log2FC)

top2de <- se.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)

DoHeatmap(se, features = top2de$gene, size = 3, angle = 0)

```

```{r, save rds file}
saveRDS(se, paste(params$data,'.rds', sep = ""))
```

